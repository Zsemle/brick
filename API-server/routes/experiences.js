const path = require('path')

const experiencesRoutes = (app, fs) => {
  const dataPath = path.join(__dirname, '..', 'data', 'experiences.json')

  const readFile = (
    callback,
    returnJson = false,
    filePath = dataPath,
    encoding = 'utf8'
  ) => {
    fs.readFile(filePath, encoding, (err, data) => {
      if (err) {
        throw err
      }

      callback(returnJson ? JSON.parse(data) : data)
    })
  }

  app.get('/experiences', (req, res) => {
    readFile((data) => {
      res.set('Access-Control-Allow-Origin', 'http://localhost:3000')
      res.send(data)
    }, true)
  })
}

module.exports = experiencesRoutes
