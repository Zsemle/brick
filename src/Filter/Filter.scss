@import '../style/typo.scss';
@import '../style/layout.scss';

.filter__toggle-button {
  display: block;
  width: 100%;
  font-size: $font-size-medium;
  line-height: $line-height-medium;
  padding: $space-small;
  cursor: pointer;
  text-align: left;
}

.filter__content {
  margin: $space-small 0;
}

.filter__check-category {
  display: block;
  font-size: $font-size-medium;
  line-height: $line-height-medium;
}

.filter__text-label,
.filter__text-input {
  display: block;
  width: 100%;
}

.filter__price-field {
  display: flex;
}
