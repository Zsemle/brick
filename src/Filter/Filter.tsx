import React, { SyntheticEvent } from 'react'
import { Experience, ExperienceCategory } from '../types/brick-types'

export interface FilterProps{
  experiences: Experience[]
  updateExperiences: (experiences:Experience[]) => void
  filterCleared: () => void
}

interface FilterState{
  isOpen: boolean
  applyTextFilter: boolean
  applyPriceFilter: boolean
  applyCategoryFilter: boolean
  filterText: string
  filterCategories: ExperienceCategory[]
  filterPriceMin: number
  filterPriceMax: number
  categoriesByText: Experience[]
  categoriesByPrice: Experience[]
  categoriesByCategory: Experience[]
}

const T:any = {
  filterToggleText: 'Filter',
  categoryFilterTitle: 'Filter by category'
}

const categories:Categories = {
  animals: 'animals',
  sports: 'sports',
  'brain games': 'brain games'
}

type Categories = {
  // eslint-disable-next-line no-unused-vars
  [thing in ExperienceCategory]: ExperienceCategory
};

class Filter extends React.Component<FilterProps, FilterState> {
  constructor (props:any) {
    super(props)
    const maxPrice = Math.max(...this.props.experiences.map(experience => experience.price.amount))
    this.state = {
      isOpen: false,
      filterText: '',
      filterPriceMin: 0,
      filterPriceMax: maxPrice,
      filterCategories: [],
      categoriesByText: [],
      categoriesByPrice: [],
      categoriesByCategory: [],
      applyTextFilter: false,
      applyPriceFilter: false,
      applyCategoryFilter: false
    }

    this.onInputTextChange = this.onInputTextChange.bind(this)
    this.onInputCheckboxChange = this.onInputCheckboxChange.bind(this)
    this.onInputPriceMinChange = this.onInputPriceMinChange.bind(this)
    this.onInputPriceMaxChange = this.onInputPriceMaxChange.bind(this)
    this.compoundFilters = this.compoundFilters.bind(this)
  }

  private onInputTextChange (e:SyntheticEvent):void{
    e.persist()
    this.setState(
      {
        filterText: (e.target as HTMLInputElement).value,
        applyTextFilter: (e.target as HTMLInputElement).value.length > 0
      },
      this.filterByText
    )
  }

  private onInputCheckboxChange (e:SyntheticEvent):void{
    e.persist()
    this.toggleFilterCategory(
      (e.target as HTMLInputElement).value as ExperienceCategory
    )
  }

  private onInputPriceMinChange (e:SyntheticEvent):void{
    e.persist()
    const newPriceLimit:number = (e.target as HTMLInputElement).value.length > 0 ? parseInt((e.target as HTMLInputElement).value) : 0
    this.setState(
      {
        filterPriceMin: newPriceLimit,
        applyPriceFilter: true
      },
      this.filterByPrice
    )
  }

  private onInputPriceMaxChange (e:SyntheticEvent):void{
    e.persist()
    const newPriceLimit:number = (e.target as HTMLInputElement).value.length > 0 ? parseInt((e.target as HTMLInputElement).value) : 0
    this.setState(
      {
        filterPriceMax: newPriceLimit,
        applyPriceFilter: true
      },
      this.filterByPrice
    )
  }

  private filterByPrice ():void{
    const {
      experiences
    } = this.props
    const {
      filterPriceMin,
      filterPriceMax
    } = this.state
    const filteredCategories:Experience[] = experiences.filter(
      experience => experience.price.amount >= filterPriceMin && experience.price.amount <= filterPriceMax
    )
    this.setState({
      categoriesByPrice: filteredCategories
    }, this.compoundFilters)
  }

  private toggleFilterCategory (category:ExperienceCategory):void{
    const { filterCategories } = this.state
    let newFilterCategories:ExperienceCategory[] = [...filterCategories]
    if (newFilterCategories.includes(category)) {
      newFilterCategories = newFilterCategories.filter(item => item !== category)
    } else {
      newFilterCategories.push(category)
    }
    this.setState({
      filterCategories: newFilterCategories,
      applyCategoryFilter: newFilterCategories.length > 0
    }, this.filterByCategories
    )
  }

  private filterByCategories ():void {
    const { filterCategories } = this.state
    const {
      experiences
    } = this.props
    const result = experiences.filter(
      (experience) => filterCategories.includes(experience.category)
    )
    this.setState({
      categoriesByCategory: result
    }, this.compoundFilters)
  }

  private filterByText ():void {
    const { filterText } = this.state
    const { experiences } = this.props
    const result = experiences.filter(
      (experience) => experience.name.toLowerCase().includes(filterText.toLowerCase())
    )
    this.setState({
      categoriesByText: result
    }, this.compoundFilters)
  }

  private compoundFilters ():void {
    const {
      updateExperiences,
      filterCleared
    } = this.props
    const {
      categoriesByText,
      categoriesByPrice,
      categoriesByCategory,
      applyTextFilter,
      applyPriceFilter,
      applyCategoryFilter
    } = this.state
    let result:Experience[] = []
    if (applyTextFilter && !applyPriceFilter && !applyCategoryFilter) {
      result = categoriesByText
    }
    if (!applyTextFilter && applyPriceFilter && !applyCategoryFilter) {
      result = categoriesByPrice
    }
    if (!applyTextFilter && !applyPriceFilter && applyCategoryFilter) {
      result = categoriesByCategory
    }
    if (applyTextFilter && applyPriceFilter && !applyCategoryFilter) {
      result = categoriesByText.filter(experience => categoriesByPrice.includes(experience))
    }
    if (applyTextFilter && !applyPriceFilter && applyCategoryFilter) {
      result = categoriesByText.filter(experience => categoriesByCategory.includes(experience))
    }
    if (!applyTextFilter && applyPriceFilter && applyCategoryFilter) {
      result = categoriesByPrice.filter(experience => categoriesByCategory.includes(experience))
    }
    if (applyTextFilter && applyPriceFilter && applyCategoryFilter) {
      const firstIntersection:Experience[] = categoriesByText.filter(experience => categoriesByPrice.includes(experience))
      result = firstIntersection.filter(experience => categoriesByCategory.includes(experience))
    }
    if (!applyTextFilter && !applyPriceFilter && !applyCategoryFilter) {
      filterCleared()
      return
    }
    updateExperiences(result)
  }

  render () {
    return (
      <div>
        <button>{T.filterToggleText}</button>
        <input type="search" value={this.state.filterText} onChange={this.onInputTextChange}/>
        <fieldset>
          <legend>{T.categoryFilterTitle}</legend>
          {Object.keys(categories).map((category, index) => (
            <label key={`${category}${index}`}>
              <input type="checkbox" onChange={this.onInputCheckboxChange} value={category}/>
              <span>{category}</span>
            </label>
          )
          )
          }
        </fieldset>
        <input min="0" type="number" value={this.state.filterPriceMin} onChange={this.onInputPriceMinChange}/>
        <input min="0" type="number" value={this.state.filterPriceMax} onChange={this.onInputPriceMaxChange}/>
      </div>
    )
  }
}

export default Filter
