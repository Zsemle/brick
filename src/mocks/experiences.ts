import { Experience } from '../types/brick-types'

const experiences:Experience[] = [
  {
    name: 'Kitten Petting',
    description: 'Pet a bunch of fluffy kittens while sipping coffee in Zwolle',
    category: 'animals',
    price: {
      amount: 10,
      currency: 'EUR'
    }
  },
  {
    name: 'Traffic Jam in Rotterdam',
    description: 'Experience the daily struggle of suburban commuters! Car not included',
    category: 'brain games',
    price: {
      amount: 42,
      currency: 'EUR'
    }
  },
  {
    name: 'An afternoon as a Peasant',
    description: 'Experience the everyday life of a 16th century Dutch farmer, living on the countryside of Zeeland. Activities include digging a hole in the mud, and peeling potatoes.',
    category: 'sports',
    price: {
      amount: 64,
      currency: 'EUR'
    }
  },
  {
    name: 'Cycling in Amersfoort',
    description: 'Join the millions of native dutchies in their most popular activity!. Rental bike included - no helmet.',
    category: 'sports',
    price: {
      amount: 30,
      currency: 'EUR'
    }
  },
  {
    name: 'Language Coaching',
    description: 'Spend an hour trying to pronounce the Dutch G without sounding like a Belgian, or worse.',
    category: 'brain games',
    price: {
      amount: 12,
      currency: 'EUR'
    }
  },
  {
    name: 'Go Shopping on a Horse',
    description: 'Come to Den Bosch to experience the dream came true.',
    category: 'animals',
    price: {
      amount: 246,
      currency: 'EUR'
    }
  },
  {
    name: 'Watch the Brexit',
    description: 'Take a seat on the beach at Zandvoort, to experience this historic moment. Chair included.',
    category: 'brain games',
    price: {
      amount: 20,
      currency: 'EUR'
    }
  },
  {
    name: 'Swimming in The Kromme Rijn at Utrecht',
    description: 'Have you ever wanted to be seen as a real weird person? Sign up now for this activity!',
    category: 'sports',
    price: {
      amount: 15,
      currency: 'EUR'
    }
  },
  {
    name: 'Escape Groeningen',
    description: 'Authentic 20th century post-apocalyptic style escape room, and a night full of activities!',
    category: 'brain games',
    price: {
      amount: 45,
      currency: 'EUR'
    }
  }
]

export default experiences
