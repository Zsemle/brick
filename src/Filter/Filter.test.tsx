import { shallow, ShallowWrapper } from 'enzyme'
import React from 'react'
import Filter, { FilterProps } from './Filter'
import experiences from '../mocks/experiences'

describe('the filter component', () => {
  const props:FilterProps = {
    experiences: experiences,
    updateExperiences: jest.fn(),
    filterCleared: jest.fn()
  }
  const wrapper:ShallowWrapper = shallow(<Filter{...props}/>)

  it('renders the component', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
