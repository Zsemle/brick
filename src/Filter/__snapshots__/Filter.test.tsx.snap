// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the filter component renders the component 1`] = `
<div>
  <button>
    Filter
  </button>
  <input
    onChange={[Function]}
    type="search"
    value=""
  />
  <fieldset>
    <legend>
      Filter by category
    </legend>
    <label
      key="animals0"
    >
      <input
        onChange={[Function]}
        type="checkbox"
        value="animals"
      />
      <span>
        animals
      </span>
    </label>
    <label
      key="sports1"
    >
      <input
        onChange={[Function]}
        type="checkbox"
        value="sports"
      />
      <span>
        sports
      </span>
    </label>
    <label
      key="brain games2"
    >
      <input
        onChange={[Function]}
        type="checkbox"
        value="brain games"
      />
      <span>
        brain games
      </span>
    </label>
  </fieldset>
  <input
    min="0"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    min="0"
    onChange={[Function]}
    type="number"
    value={246}
  />
</div>
`;
